sympy/diffgeom/diffgeom.py

```diff
@@ -1084,6 +1086,377 @@ def __call__(self, field):
+        self.order = self.covariant_order+self.contravariant_order
+        indices = list(itertools.product(range(self.n),repeat=self.order))
+        self.contravariant_slots=self._contravariant_slots(T)
+        self.covariant_slots=list(range(self.order))
+        [self.covariant_slots.remove(i) for i in self.contravariant_slots]
+        def basishelper(p,k):
+            if self.contravariant_slots.count(k) == 0:
+                return self.base_vectors[p[k]]
+            else:
+                return self.coords[p[k]]
+        Tbases = [
+            (p,[basishelper(list(p),k) for k in range(self.order)]) for p in indices
+        ]
+        if self.order==0: #scalar case
+            coeffs = [T]
+        else:
+            coeffs = []
+            for (p,exI) in Tbases:
+                Tval = T.rcall(*exI)
+                if Tval != 0:
+                    coeffs.append(Tval)
+                else:
+                    indices.remove(p)
+        self.tensor = dict(zip(indices,coeffs))
+
+    def to_NDimArray(self):
+        """
+        Converts from internal sparse dict representation to sympy.tensor.MutableSparseNDimArray
+
+        Example
+        =======
+        >>> from sympy.diffgeom import TensorProduct, Manifold, Patch, CoordSystem, TensorArray
+        >>> from sympy import Function
+        >>> from sympy.functions import sin
+        >>> M = Manifold('Reisner-Nordstrom', 4)
+        >>> p = Patch('origin', M)
+        >>> cs = CoordSystem('spherical', p, ['t', 'r', 'theta', 'phi'])
+        >>> t, r, theta, phi = cs.coord_functions()
+        >>> dt, dr, dtheta, dphi = cs.base_oneforms()
+        >>> f=Function('f')
+        >>> TP=TensorProduct
+        >>> metric = f(r)**2*TP(dt, dt) - f(r)**(-2)*TP(dr, dr) - r**2*TP(dtheta, dtheta) - r**2*sin(theta)**2*TP(dphi, dphi)
+        >>> TensorArray(metric).to_NDimArray()
+        [[f(r)**2, 0, 0, 0], [0, -1/f(r)**2, 0, 0], [0, 0, -r**2, 0], [0, 0, 0, -r**2*sin(theta)**2]]
+        """
+        a = MutableSparseNDimArray.zeros(*tuple(self.n for _ in range(self.order)))
+        for (index,coeffs) in list(self.tensor.items()):
+            a[index] = coeffs
+        return a
+
+    def _init_from_array(self,component_array,variance,coordinate_system):
+        self.CoordSystem = coordinate_system
+        self.coords = self.CoordSystem.coord_functions()
+        self.base_oneforms = self.CoordSystem.base_oneforms()
+        self.base_vectors = self.CoordSystem.base_vectors()
+        self.n = len(self.coords)
+        self.order = len(variance)
+        self.covariant_slots=[k for k in range(self.order) if variance[k]==1]
+        self.covariant_order = len(self.covariant_slots)
+        self.contravariant_slots=[k for k in range(self.order) if variance[k]==-1]
+        self.contravariant_order = len(self.contravariant_slots)
+        indices = list(itertools.product(range(self.n),repeat=self.order))
+        self.tensor = dict(zip(indices,component_array))
```

Comment:
It looks like all index combinations are included. A dense representation?

Reply:
If it's initialized from an array, yes.  Mostly the class is useful when it is initialized with a diffgeom object, because this is a useless library otherwise.

Reply:
Which would be the principal way of initialization?

Reply:
Two examples: 

```Python
>>> from sympy.diffgeom import (Manifold, Patch, CoordSystem, TensorProduct,  expand_tensor)
>>> TP = TensorProduct
>>> M = Manifold('Reisner-Nordstrom', 4)
>>> p = Patch('origin', M)
>>> cs = CoordSystem('spherical', p, ['t', 'r', 'theta', 'phi'])
>>> t, r, theta, phi = cs.coord_functions()
>>> dt, dr, dtheta, dphi = cs.base_oneforms()
>>> TP(dt + dr, dt + dr)
TensorProduct(dr + dt, dr + dt)
>>> expand_tensor(TP(dt+dr,dt+dr))  #This uses the class to expand the tensor
TensorProduct(dr, dr) + TensorProduct(dr, dt) + TensorProduct(dt, dr) + TensorProduct(dt, dt)

>>> from sympy.diffgeom import (metric_to_Christoffel_2nd, TensorArray)
>>> from sympy import (Function, sin)
>>> f=Function('f')
>>> metric = f(r)**2*TP(dt, dt) - f(r)**(-2)*TP(dr, dr) - r**2*TP(dtheta, dtheta) - r**2*sin(theta)**2*TP(dphi, dphi)
>>> ch_2nd = metric_to_Christoffel_2nd(metric)
>>> G=TensorArray(metric)
>>> assert(G.covD(ch_2nd).to_tensor() == 0)  #the covariant derivative as a (3,0) tensor!
```